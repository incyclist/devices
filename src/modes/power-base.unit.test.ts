import DeviceAdapterBase from '../Device';
import DeviceProtocolBase from '../DeviceProtocol';
import PowerMeterMode from './power-base'

class MockAdapter extends DeviceAdapterBase {
    constructor() {
        super( new DeviceProtocolBase())
    }
}

describe('PowerMeterMode', () => {

    let mode;
    beforeAll( ()=> {
        mode = new PowerMeterMode( new MockAdapter());
    })

    describe ( 'getWeight', () => { 
        const DEFAULT_WEIGHT = 85;

        describe ( 'adapter has getWeight()', () => { 
            let testWeight;
            beforeAll(()=> {
                mode.adapter.getWeight = ()=> testWeight;
            })

            afterAll(()=> {
                mode.adapter.getWeight = undefined;
            })

            test('should return the weight from adapter', () => {
                testWeight = 100;
                expect(mode.getWeight()).toBe(testWeight);
            })
            test('should return default if adapter.getWeight() returns undefined or 0', () => { 
                testWeight = 0;
                expect(mode.getWeight()).toBe(DEFAULT_WEIGHT);
                testWeight = undefined;
                expect(mode.getWeight()).toBe(DEFAULT_WEIGHT);
            })
            
            

        })
        describe ( 'adapter does not have getWeight()', () => { 
            test('should return default weight', () => {
                expect(mode.getWeight()).toBe(DEFAULT_WEIGHT);
            })

        })

    })

    describe ( 'calculateSpeedAndDistance', () => { 

        beforeEach( ()=> {
            mode = new PowerMeterMode( new MockAdapter());
        })


        test('150W,0%, 70kg', ()=>{
            const values = [];
            for (let i=0; i<30; i++) {
                values.push({t:i+1, ...mode.calculateSpeedAndDistance(150, 0,78.5, 1)});
            }
            expect(values).toMatchSnapshot()
            //expect(values[9].speed).toEqual(18);
        })

        test('0W, 45km/h, 0%', ()=>{
            mode.data.speed = 45;
            const values = [];
            for (let i=0; i<30; i++) {
                values.push({t:i+1, ...mode.calculateSpeedAndDistance(0, 0,80, 1)});
            }
            expect(values).toMatchSnapshot()

        })


        test('0W, 30km/h, -7%, 72kg', ()=>{
            // maximum speed that can be achieved with 0W and -7% is 52km/h
            // loop could be exended to >1000 but that would not change the result
            mode.data.speed = 45;
            const values = [];
            for (let i=0; i<100; i++) {
                values.push({t:i+1, ...mode.calculateSpeedAndDistance(0, -7,72, 1,{cwA:0.3784, cRR:0.005, rho:1.18})});
            }
            expect(values[values.length-1].speed).toBeCloseTo(52,0)

        })

    })
})