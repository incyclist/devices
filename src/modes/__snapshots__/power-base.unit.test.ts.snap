// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PowerMeterMode calculateSpeedAndDistance 0W, 45km/h, 0% 1`] = `
Array [
  Object {
    "distance": 12.044852056804704,
    "speed": 43.361467404496935,
    "t": 1,
  },
  Object {
    "distance": 11.619920378007851,
    "speed": 41.83171336082827,
    "t": 2,
  },
  Object {
    "distance": 11.22216353417546,
    "speed": 40.399788723031655,
    "t": 3,
  },
  Object {
    "distance": 10.848939286871508,
    "speed": 39.056181432737425,
    "t": 4,
  },
  Object {
    "distance": 10.497940607233762,
    "speed": 37.79258618604155,
    "t": 5,
  },
  Object {
    "distance": 10.167143700914984,
    "speed": 36.60171732329395,
    "t": 6,
  },
  Object {
    "distance": 9.854765476120448,
    "speed": 35.477155714033614,
    "t": 7,
  },
  Object {
    "distance": 9.55922850123546,
    "speed": 34.413222604447654,
    "t": 8,
  },
  Object {
    "distance": 9.27913194919597,
    "speed": 33.40487501710549,
    "t": 9,
  },
  Object {
    "distance": 9.01322736229164,
    "speed": 32.44761850424991,
    "t": 10,
  },
  Object {
    "distance": 8.760398324840484,
    "speed": 31.53743396942574,
    "t": 11,
  },
  Object {
    "distance": 8.519643324235165,
    "speed": 30.670715967246597,
    "t": 12,
  },
  Object {
    "distance": 8.290061228993732,
    "speed": 29.844220424377436,
    "t": 13,
  },
  Object {
    "distance": 8.070838927011604,
    "speed": 29.055020137241772,
    "t": 14,
  },
  Object {
    "distance": 7.8612407564773035,
    "speed": 28.300466723318294,
    "t": 15,
  },
  Object {
    "distance": 7.660599431958302,
    "speed": 27.57815795504989,
    "t": 16,
  },
  Object {
    "distance": 7.468308223490239,
    "speed": 26.885909604564862,
    "t": 17,
  },
  Object {
    "distance": 7.283814190479876,
    "speed": 26.221731085727555,
    "t": 18,
  },
  Object {
    "distance": 7.1066123073958885,
    "speed": 25.5838043066252,
    "t": 19,
  },
  Object {
    "distance": 6.936240346497176,
    "speed": 24.970465247389832,
    "t": 20,
  },
  Object {
    "distance": 6.772274405707061,
    "speed": 24.38018786054542,
    "t": 21,
  },
  Object {
    "distance": 6.614324988315657,
    "speed": 23.811569957936364,
    "t": 22,
  },
  Object {
    "distance": 6.462033556358203,
    "speed": 23.26332080288953,
    "t": 23,
  },
  Object {
    "distance": 6.315069491957307,
    "speed": 22.734250171046305,
    "t": 24,
  },
  Object {
    "distance": 6.173127411166519,
    "speed": 22.223258680199468,
    "t": 25,
  },
  Object {
    "distance": 6.035924783333051,
    "speed": 21.729329219998984,
    "t": 26,
  },
  Object {
    "distance": 5.903199816042478,
    "speed": 21.25151933775292,
    "t": 27,
  },
  Object {
    "distance": 5.774709571583408,
    "speed": 20.788954457700267,
    "t": 28,
  },
  Object {
    "distance": 5.65022828578799,
    "speed": 20.340821828836763,
    "t": 29,
  },
  Object {
    "distance": 5.529545864235205,
    "speed": 19.90636511124674,
    "t": 30,
  },
]
`;

exports[`PowerMeterMode calculateSpeedAndDistance 150W,0%, 70kg 1`] = `
Array [
  Object {
    "distance": 1.9549056373532239,
    "speed": 7.037660294471606,
    "t": 1,
  },
  Object {
    "distance": 2.7322466363557876,
    "speed": 9.836087890880835,
    "t": 2,
  },
  Object {
    "distance": 3.3142733660764017,
    "speed": 11.931384117875046,
    "t": 3,
  },
  Object {
    "distance": 3.791649320273353,
    "speed": 13.649937552984072,
    "t": 4,
  },
  Object {
    "distance": 4.199749085719485,
    "speed": 15.119096708590146,
    "t": 5,
  },
  Object {
    "distance": 4.556899778809862,
    "speed": 16.404839203715504,
    "t": 6,
  },
  Object {
    "distance": 4.874139102249889,
    "speed": 17.5469007680996,
    "t": 7,
  },
  Object {
    "distance": 5.158785987561405,
    "speed": 18.571629555221058,
    "t": 8,
  },
  Object {
    "distance": 5.416033317429446,
    "speed": 19.497719942746006,
    "t": 9,
  },
  Object {
    "distance": 5.64975618401961,
    "speed": 20.339122262470596,
    "t": 10,
  },
  Object {
    "distance": 5.862961466538358,
    "speed": 21.106661279538088,
    "t": 11,
  },
  Object {
    "distance": 6.058055776484246,
    "speed": 21.809000795343287,
    "t": 12,
  },
  Object {
    "distance": 6.237014032714115,
    "speed": 22.453250517770815,
    "t": 13,
  },
  Object {
    "distance": 6.401490267225212,
    "speed": 23.045364962010762,
    "t": 14,
  },
  Object {
    "distance": 6.55289315820542,
    "speed": 23.590415369539514,
    "t": 15,
  },
  Object {
    "distance": 6.6924391364969775,
    "speed": 24.09278089138912,
    "t": 16,
  },
  Object {
    "distance": 6.821190739502056,
    "speed": 24.556286662207402,
    "t": 17,
  },
  Object {
    "distance": 6.940084974346455,
    "speed": 24.984305907647236,
    "t": 18,
  },
  Object {
    "distance": 7.049954742392615,
    "speed": 25.379837072613416,
    "t": 19,
  },
  Object {
    "distance": 7.151545336812722,
    "speed": 25.745563212525802,
    "t": 20,
  },
  Object {
    "distance": 7.245527371891223,
    "speed": 26.083898538808402,
    "t": 21,
  },
  Object {
    "distance": 7.332507081544413,
    "speed": 26.397025493559887,
    "t": 22,
  },
  Object {
    "distance": 7.413034646339645,
    "speed": 26.68692472682272,
    "t": 23,
  },
  Object {
    "distance": 7.487611020606611,
    "speed": 26.9553996741838,
    "t": 24,
  },
  Object {
    "distance": 7.55669360220223,
    "speed": 27.20409696792803,
    "t": 25,
  },
  Object {
    "distance": 7.620700997277949,
    "speed": 27.43452359020062,
    "t": 26,
  },
  Object {
    "distance": 7.680017068361012,
    "speed": 27.648061446099646,
    "t": 27,
  },
  Object {
    "distance": 7.734994407975727,
    "speed": 27.845979868712618,
    "t": 28,
  },
  Object {
    "distance": 7.785957346446739,
    "speed": 28.029446447208258,
    "t": 29,
  },
  Object {
    "distance": 7.833204577770267,
    "speed": 28.19953647997296,
    "t": 30,
  },
]
`;
